### ISSUE TITLE
Allow CORS when serving (WebDAV)

### ISSUE THREAD

**Issue by alensiljak** [rclone/rclone#5078]
#### What is your current rclone version (output from `rclone version`)?

rclone v1.54.0
- os/arch: linux/amd64
- go version: go1.15.7

I also use it through Termux on Android

rclone v1.54.0-DEV
- os/arch: android/arm64
- go version: go1.15.6

or through Arch installation in Termux.

#### What problem are you are trying to solve?

Working with remote Org note files over WebDAV.

Long version:

To expand a little bit - I am slowly migrating from native mobile apps to PWAs and, in parallel, my info collection (notes, tasks, etc.) to Org mode. While Orgzly, a native Android app, is great, I believe that PWAs offer better opportunities, both to the developers and to the users. Therefore, I would like to use a PWA (https://organice.200ok.ch/) for the purpose of working with .org files.
This app supports working over WebDAV (since local file access in PWAs is still clunky). Unfortunately, my storage provider, while supporting WebDAV access, does not support CORS. Therefore, it is impossible to access the storage via WebDAV in any JavaScript-based browser application.
I was hoping rclone would come to the rescue with it's `rclone serve webdav` but unfortunately that provides exactly the same result - it doesn't support CORS. Or at least I can't find one in the docs and help texts.

#### How do you think rclone should be changed to solve that?

Include CORS header `Access-Control-Allow-Origin: *` when serving files over WebDAV (and potentially other protocols).

Similar to #2575, I would really appreciate if CORS headers were supported (through an option parameter) when serving files, either via HTTP or WebDAV (`rclone server webdav`).

**Comment by darthShadow** [comment]
You should be able to get it working by adding the headers manually as needed for now via the `--headers` flag: https://rclone.org/docs/#header

**Comment by MisterY** [comment]
Hmm... I'm trying it out and it does not seem to work. It seems to me that this is meant for the Request headers, when rclone is accessing the remote service?
Just to be clear, I'm looking for the server-side headers "Access-Control-Allow-Headers: *", when rclone is used via `rclone serve webdav`.

What I'm getting from rclone in that case is:

![image](https://user-images.githubusercontent.com/462445/109427786-46b23f80-79f4-11eb-8510-861991a36c6c.png)

There's no CORS header.
The command I use is:

`rclone serve webdav xyz:notes/org --vfs-cache-mode full --header "Access-Control-Allow-Headers: *"`

Perhaps I'm doing something wrong here? Thanks for the assistance!

**Comment by MisterY** [comment]
I also tried adding the `Access-Control-Allow-Origin: *` but to no avail.

**Comment by darthShadow** [comment]
On a more careful look, yeah, this method won't work for the serve command since that uses its own http handler which doesn't have support for adding custom headers.

**Comment by Navdevl** [comment]
@darthShadow Can I pick up and work on this? I am trying to slowly step in one by one and understand the codebase. This seems like a place where I don't have to tackle a lot of complications and can figure out easily. Do you think I can work on this?

**Comment by darthShadow** [comment]
You can definitely give it a try.

You will need to look at the following files to get an idea of how rclone handles adding arbitrary headers:

https://github.com/rclone/rclone/blob/master/fs/config/configflags/configflags.go -> For parsing
https://github.com/rclone/rclone/blob/master/fs/fshttp/http.go -> For adding

And then add similar logic for https://github.com/rclone/rclone/blob/master/cmd/serve/httplib/httplib.go

**Comment by Navdevl** [comment]
@darthShadow thank you for the directions. I'll work on this over the weekend and quickly raise a PR on the same.

**Comment by Navdevl** [comment]
@darthShadow @MisterY - Sorry, I might not be able to work on this for next couple weeks. If somebody is interested to pick this up, let them. If this is still open after 2 weeks, I can definitely work on this. Sorry. :/

**Comment by uGeek** [comment]
Are you going to continue with this project? There are different clients (KeeWeb, diffuse,...) that use CORS to function and they give error.

Thanks

**Comment by Arnie97** [comment]
> via the `--headers` flag

> this method won't work for the serve command since that uses its own http handler which doesn't have support for adding custom headers

Here I propose a new flag that takes care of headers in HTTP server responses, just like what `--headers` did for HTTP client requests.  I'm not sure what would be the proper name for the new flag though.  Maybe `--headers-serve`?

**Comment by yuudi** [comment]
I have another issue with CORS when working with `rclone rcd`, and I found out the problem

The custom headers only apply **AFTER** the authorization is accepted, however when browser check CORS, they don't take `Authorization` header due to security concerns.

So, the best way is to use another option (not exposed to user) to set http header before the `Auth` middleware

I found this code

https://github.com/rclone/rclone/blob/4f8dab8bcec40f34f02074549747437ac17b0bef/lib/http/server.go#L229

I think a new middleware (maybe `HeadersPreAuth`) before this line will help

And the old cli option `--rc-allow-origin` can be combined to a new option to control both `rclone rcd` and `rclone serve` to solve CORS and other http header issue

**Comment by yuudi** [comment]
> 
> The thing to do would be to move the `allow-origin` option into `lib/http` then all the servers `rc` and other can make use of it.
> 
> Want to have a go at that @yuudi ?

I'll give it a try this week

**Comment by alensiljak** [comment]
Related PR: https://github.com/rclone/rclone/pull/7089

**Comment by nitincodery** [comment]
So how do you serve phone files to webdav with CORS? what is the command to run in termux to serve files? I tried:

`rclone serve webdav --allow-origin "Access-Control-Allow-Origin: *" directory/`

Still unable to access the directory in organice, if its fixed, what command to run?

**Comment by alensiljak** [comment]
Hi @nitincodery ! No, unfortunately it doesn't work for me, either. The issue is a bit more nuanced now since I'm getting an error about a missing command or something like that so I blamed my setup. My client is sending Fetch and that is reported as being unsupported although it is listed right there in the allowed commands, in the response header from rclone.
So the short answer is no. The longer one is that the CORS header is there but my scenario still doesn't work. And obviously yours, too. I just didn't get to trying out with other clients or commands.

**Comment by nitincodery** [comment]
Yes, I was also searching about PROPFIND error, I thought CORS is fixed but now this might be something else, meanwhile, what I found is there's an organice fork with android app to provide local access. It is working for my org files synced to my phone via syncthing, so I don't think I'd bang more head with rclone webdav on localhost.

https://github.com/ieugen/organice/releases/tag/android-local-fs.v1

**Comment by alensiljak** [comment]
@nitincodery, thanks for that link! I have several other use cases where I'd like to access my own files at a cloud provider since CORS doesn't let me do that from the browser (but it does from a console application, unfair!).
However, it is not high priority at the moment.

**Comment by nitincodery** [comment]
Also, I just found you made cashier, and I also use ledger-cli heavily to manage my personal expenses as well my company's, I was struggling to use ledger cli on phone as org files, I've also made a org table through which I due my entries into ledger, which was made because of atleast get read view of entries on phone via orgro app.

But I just found about cashier by you, I'm gonna try it out and it looks promising, will it too won't work due to CORS mess? or it does work? I'm just reading about its rust cashier-server app to run inside termux, I hope it works, and I know rust so I'll try to contribute to it some too.

**Comment by alensiljak** [comment]
@nitincodery, oh, thanks a lot! It definitely does work since that's what I'm using for years now. If you run into any issues, please file them in my repo. Currently I'm probably the only one using it so it is hard to see things from somebody else's perspective.

**Comment by ST-Saint** [comment]
I also had problems with WebDAV CORS. I got the following response
```http
HTTP/1.1 200 OK
Accept-Ranges: bytes
Access-Control-Allow-Headers: authorization, Content-Type
Access-Control-Allow-Origin: *
Access-Control-Request-Method: POST, OPTIONS, GET, HEAD
Allow: OPTIONS, LOCK, DELETE, PROPPATCH, COPY, MOVE, UNLOCK, PROPFIND
Dav: 1, 2
Ms-Author-Via: DAV
Server: rclone/v1.64.2
Date: Mon, 06 Nov 2023 05:29:53 GMT
Content-Length: 0
```
and browser reports `Response body is not available to scripts (Reason: CORS Method Not Found)`.

Not quite familiar with CORS but should the `Access-Control-Request-Method` in the response be `Access-Control-Allow-Method`?

**Comment by viocha** [comment]
The `access-control-max-age:86400` header is missing, which prevents OPTIONS requests from being cached, resulting in repeated OPTIONS requests.

**Comment by viocha** [comment]
<img width="1355" height="232" alt="Image" src="https://github.com/user-attachments/assets/1511cda7-a314-40f9-86c1-cd058b278869" />

**Comment by ncw** [comment]
@viocha do you want to send a PR to fix? Or if you don't want to fix, please make a new issue - thank you

**Comment by viocha** [comment]
> [@viocha](https://github.com/viocha) do you want to send a PR to fix? Or if you don't want to fix, please make a new issue - thank you

Sure, I can try to send a PR to fix this. It looks straightforward.
